visibilityChanged.js

```diff
@@ -1,35 +1,84 @@
-        frequency: 100,
-        previousVisibility : null
-    };
-
-    var methods = {};
-    methods.checkVisibility = function (element, options) {
-        if (jQuery.contains(document, element[0])) {
-            var previousVisibility = options.previousVisibility;
-            var isVisible = element.is(':visible');
-            options.previousVisibility = isVisible;
-            if (previousVisibility == null) {
-                if (options.runOnLoad) {
-                    options.callback(element, isVisible);
-                }
-            } else if (previousVisibility !== isVisible) {
-                options.callback(element, isVisible);
-            }
-
-            setTimeout(function() {
-                methods.checkVisibility(element, options);
-            }, options.frequency);
-        } 
-    };
-
-    $.fn.visibilityChanged = function (options) {
-        var settings = $.extend({}, defaults, options);
-        return this.each(function () {
-            methods.checkVisibility($(this), settings);
-        });
-    };
+    var
+		checkVisibility = function(element, options) {
+			var
+				obj = {
+					defaultOptions: {
+						changed: function (element, currentVisibility, previousVisibility, event) {},
+						same: function(element, visibility, event) {},
+						eventList: [
+							'click',
+							'mouseMove',
+							'keyUp',
+							'scroll',
+							'resize'
+						]
+					},
+					options: {},
+					element: null,
+					check: function(e) {
+						var t = this;
+						
+						t.element.data('previousVisibility', t.element.data('currentVisibility'));
+						
+						t.element.data('currentVisibility', t.element.is(':visible'));
+						
+						if (t.element.data('previousVisibility') == t.element.data('currentVisibility')) {
+							t.options.same(t.element, t.element.data('currentVisibility'), e);
+						} else {
+							t.options.changed(t.element, t.element.data('currentVisibility'), t.element.data('previousVisibility'), e);
+						}
+						
+						return t;
+					},
+					init: function(element, options) {
+						var t = this;
+						t.options = $.extend({}, t.defaultOptions, options);
+						t.element = $(element);
+						
+						if (t.element.data('currentVisibility')) {
+							// Do nothing.
+						} else {
+							t.element.data('currentVisibility',  t.element.is(':visible'));
+							t.element.data('previousVisibility', null);
+						}
+						
+						if (t.options.eventList) {
+							$.each(t.options.eventList, function(i, ev) {
+								$(document).on(
+									ev,
+									function(e) {
+										t.check(e);
+									}
+								);
+							});
+						}
+						
+						return t;
+					}
+				}
+			;
+			obj.init(element, options);
+			
+			return obj;
+		}
+	;
+	
+	$.fn.extend({
+		'visibilityChanged': function(options) {
+			var options = options || {};
+			return new checkVisibility(this, options);
+		}
+	});
+	
+	$(document).ready(function() {
+		$('*.watchVisibility').each(function(i, el) {
```

Comment:
Why is the selection not just `.watchVisibility`? Seems like the `*` isn't required and would spoil any sizzle optimizations.